### ISSUE TITLE
Turning to Source when a widget is focused adds empty paragraphs around it

### ISSUE THREAD

**Issue by dryoma** [w8tcha/CKEditor-TextSelection-Plugin#28]
Hello,

## Steps to reporduce:

1. Create a widget plugin.
2. Enter the editor, add a widget to an arbitrary position.
    We will have a structure like this:
    ```html
    <p>Some text... </p>
    
    <figure class="widget-name">
      <img .. some attrs />
      <figcaption>edit me...</figcaption>
    </figure>
    
    <p>Some other text.</p>
    ```
3. Click on the widget - on the widget itself, not on any of the editables - thus making it focused.
4. Click Source toolbar button.

### Results in:

Empty paragraphs are inserted before and after the widget:
```html
<p>Some text... </p>

<p>&nbsp;</p>

<figure class="widget-name">
  <img .. some attrs />
  <figcaption>edit me...</figcaption>
</figure>

<p>&nbsp;</p>

<p>Some other text.</p>
```

## Other details

* This does not happen if a widget's editable is focused prior to switching to Source, only if the widget as a whole is focused.
* It seems to not depend on the `fillEmptyBlocks` or `enterMode` settings. I tried to make `<p>`a root widget element (replacing figcaption with span) - the result was the same, empty paragraphs. Whether there is an `&nbsp;` seems irrelevant.

After some debugging I found out that it's the [`createBookmark()`](https://github.com/ckeditor/ckeditor-dev/blob/3b7f188ce4222fb0795674e052b4117caa87dc40/core/dom/range.js#L709) method called [here](https://github.com/w8tcha/CKEditor-TextSelection-Plugin/blob/master/textselection/plugin.js#L106) that adds `<span data-cke-bookmark="1" id="cke_bm_699S" style="display: none;">&nbsp;</span>` before the widget (and after the p with `Some text... `) and a similar one after it. Which are later enclosed in paragraphs, then removed, but the paragraphs remain.

Thanks!

------------
The code of the widget:
```js
CKEDITOR.plugins.add('img-figure', {
  requires: 'widget',

  icons: 'img-figure',

  init: function (editor) {
    editor.widgets.add('img-figure', {
      button: 'Create a img-figure',

      template:
        '<figure class="Picture">' + 
          '<image class="Picture-img" src="http://docs.ckeditor.com/resources/images/class-m.png">' +
          '<figcaption class="Picture-imgCaption"></figcaption>' +
        '</figure>',

      editables: {
        caption: {
          selector: '.Picture-imgCaption',
          allowedContent: 'strong em'
        }
      },

      allowedContent:
        'figure(Picture),img(Picture-img),figcaption(Picture-imgCaption)',

      requiredContent: 'figure(Picture)',

      upcast: function (element) {
        return element.name === 'figure' &&
          element.hasClass('Picture');
      },

      init: function () { },
    });
  },
});
```

**Comment by dryoma** [comment]
An update: I created a ticket in CKEditor to discuss this issue and ask that maybe they'd be willing to do something about it, just in case. According to one of the devs:

> About `createBookmark`, it's clearly described that this method modify DOM elements inside editor. If you are using it, you should be aware that such elements will be added to editor. And because of that, you should handle removing them from editor.
> We cannot change way how `createBookmark` work, because this behaviour is used in many cases.

**Comment by w8tcha** [comment]
This is one of the limitations of the plugin. Unless i found a better solution to save the position this issue will never be fixed

**Comment by w8tcha** [comment]
i'm testing a new solution

https://github.com/w8tcha/CKEditor-TextSelection-Plugin/releases/tag/v1.07

**Comment by dryoma** [comment]
Hi and thanks for addressing the issue! Sorry to have missed that. The selection seems to work, however there are some other issues in the master tree (will put toghether a separate issue).
